tree/ntuple/v7/inc/ROOT/RDaos.hxx

```diff
@@ -0,0 +1,267 @@
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#ifndef ROOT7_RDaos
+#define ROOT7_RDaos
+
+#include <ROOT/RStringView.hxx>
+
+#include <daos.h>
+#include <gurt/common.h>
+// Undefine these macros in 'gurt/common.h' as they conflict with std::min()/std::max().
+// TODO: report this to DAOS/gurt maintainers.
+#undef min
+#undef max
+
+#include <numeric>
+#include <functional>
+#include <type_traits>
+#include <vector>
+#include <memory>
+
+namespace ROOT {
+
+namespace Experimental {
+namespace Detail {
+class RDaosContainer;
+
+/**
+  \class RDaosPool
+  \brief A RDaosPool provides access to containers in a specific DAOS pool.
+  */
+class RDaosPool {
+   friend class RDaosContainer;
+private:
+   daos_handle_t fPoolHandle{};
+   daos_pool_info_t fPoolInfo{};
+   uuid_t fPoolUuid{};
+
+public:
+   RDaosPool() = delete;
+   RDaosPool(std::string_view poolUuid, std::string_view serviceReplicas);
+   ~RDaosPool();
+};
+
+/**
+  \class RDaosObject
+  \brief Provides low-level access to DAOS objects in a container.
+  */
+template <typename T_dkey, typename T_akey>
+class RDaosObject {
+private:
+   daos_handle_t fObjectHandle;
+
+   template <typename... Ts> struct __is_container {};
+   template <typename T, typename _ = void>
+   struct is_container : std::false_type {};
+   template <typename T>
+   struct is_container<T, std::conditional_t<false,
+                                            __is_container<typename T::pointer,
+                                                           decltype(std::declval<T>().size()),
+                                                           decltype(std::declval<T>().data())>, void>
+                       > : std::true_type {};
+
+   // Provide support for UINT64 dkey/akey.
+   template <typename T, std::enable_if_t<!is_container<T>::value, int> = 0>
+   static size_t key_size(T& x) { return sizeof(x); }
+   template <typename T, std::enable_if_t<!is_container<T>::value, int> = 0>
+   static typename std::add_pointer<T>::type key_data(T& x) { return &x; }
+
+   // Provide support for std::string/std::vector dkey/akey.
+   template <typename T, std::enable_if_t<is_container<T>::value, int> = 0>
+   static size_t key_size(T& x) { return x.size(); }
+   template <typename T, std::enable_if_t<is_container<T>::value, int> = 0>
+   static typename T::pointer key_data(T& x) { return const_cast<typename T::pointer>(x.data()); }
+
+public:
+   using tDistributionKey = T_dkey;
+   using tAttributeKey = T_akey;
+  
+   /// \brief Contains required information for a single fetch/update operation.
+   struct FetchUpdateArgs {
```

Comment:
Seem a little bulky for a header.  Can it go into the translation unit?

Reply:
This relies on template paramenters.  For this (and other declarations), I think I will move from the template definition on the header to the translation unit + explicit instantiation for the `uint64_t` type.